### PR TITLE
test: speed up wallet_avoidreuse, add logging

### PR DESCRIPTION
**Pull request by jonatack**
Inspired by PRs #17340 and #15881.

- add logging
- pass -whitelist in `set_test_params` to speed up transaction relay

`wallet_avoidreuse.py` is not intended to test P2P transaction relay/timing, so it should be fine to do this here. This reduces test run time variability and speeds up the test by 2-3 times on average.

Test run times in seconds:

- before: 20, 24, 22, 17, 27, 40, 30

- after: 10, 10, 8, 9, 10, 7, 8

### CODE REVIEW

**Comment by jonatack** [comment]
Test output with logging:
```
$ test/functional/wallet_avoidreuse.py 
2019-11-04T09:25:24.487000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_5geycz8_
2019-11-04T09:25:25.657000Z TestFramework (INFO): Test wallet files persist the avoid_reuse flag
2019-11-04T09:25:26.715000Z TestFramework (INFO): Test immutable wallet flags
2019-11-04T09:25:27.637000Z TestFramework (INFO): Test fund send fund send dirty
2019-11-04T09:25:31.799000Z TestFramework (INFO): Test fund send fund send
2019-11-04T09:25:34.649000Z TestFramework (INFO): Stopping nodes
2019-11-04T09:25:34.955000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_5geycz8_ on exit
2019-11-04T09:25:34.955000Z TestFramework (INFO): Tests successful
```

File: test/functional/wallet_avoidreuse.py [hunk 1]
```diff
@@ -67,6 +67,9 @@ class AvoidReuseTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = False
         self.num_nodes = 2
+        # This test isn't testing txn relay/timing, so set whitelist on the
+        # peers for instant txn relay. This speeds up the test run time 2-3x.
+        self.extra_args = [['-whitelist=127.0.01']] * 2
```
Comment (practicalswift) [inline test/functional/wallet_avoidreuse.py]: `127.0.0.1`? :)

Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Thanks @practicalswift, good eye :)

Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Fixed.

File: test/functional/wallet_avoidreuse.py [hunk 2]
```diff
@@ -58,78 +58,85 @@ def assert_unspent(node, total_count=None, total_sum=None, reused_supported=None
...
+        self.test_fund_send_fund_send_dirty()
         reset_balance(self.nodes[1], self.nodes[0].getnewaddress())
         self.test_fund_send_fund_send()
 
     def test_persistence(self):
-        '''Test that wallet files persist the avoid_reuse flag.'''
-        # Configure node 1 to use avoid_reuse
-        self.nodes[1].setwalletflag('avoid_reuse')
+        """Test that wallet files persist the avoid_reuse flag."""
+        self.log.info("Test wallet files persist avoid_reuse flag")
```
Comment (jachiang) [inline test/functional/wallet_avoidreuse.py]: s/Test wallet files/Test that wallet files

**Review by jachiang — CHANGES_REQUESTED** [review]
Thanks @jonatack!

tACK d2edf02d54ac3286f721ee7d637f5c5e886fbd34 
Reviewed test, also don't see any need for tx propagation delay.

My informal run times confirm improvement.
Before 463eab5e : 31.9 / 22.9 / 35.7 / 29.8 / 34.8 / 29.7 / 
With patch d2edf02: 14.7 / 11.9 / 10.7/ 12.9 / 13.7 / 13.7 /12.4

However, I am unsure of the source of variability. I don't see an improvement there.

File: test/functional/wallet_avoidreuse.py [hunk 3]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env python3
-# Copyright (c) 2018 The Bitcoin Core developers
+# Copyright (c) 2018-2019 The Bitcoin Core developers
```
Comment (Sjors) [inline test/functional/wallet_avoidreuse.py]: Generally don't touch this: maintainers run a script once a year to update this for all files.

**Review by Sjors — APPROVED** [review]
ACK d2edf02 on macOS 10.15 with `--enable-debug` (which is slower, but better at catching issues).

Before: 18-25 seconds
After: 10 seconds

Part of inconsistency can come from caching.

File: test/functional/wallet_avoidreuse.py [hunk 4]
```diff
@@ -191,45 +199,45 @@ def test_fund_send_fund_send(self):
...
 
         self.nodes[0].sendtoaddress(fundaddr, 10)
         self.nodes[0].generate(1)
         self.sync_all()
 
-        # listunspent should show 2 total outputs (5, 10 btc), one unused (5), one reused (10)
+        # listunspent should show 2 total outputs (5, 10 btc), one unused (5), one reused (10).
         assert_unspent(self.nodes[1], total_count=2, total_sum=15, reused_count=1, reused_sum=10)
-        # getbalances should show 10 used, 5 btc trusted
+        # getbalances should show 10 used, 5 btc trusted.
```
Comment (MarcoFalke) [inline test/functional/wallet_avoidreuse.py]: I don't know if it makes sense to add trailing dots to each comment that is just a single sentence

File: test/functional/wallet_avoidreuse.py [hunk 5]
```diff
@@ -191,45 +199,45 @@ def test_fund_send_fund_send(self):
...
+        # getbalances should show 10 used, 1 btc trusted.
         assert_balances(self.nodes[1], mine={"used": 10, "trusted": 1})
 
-        # node 1 should now have about 1 btc left (no dirty) and 11 (including dirty)
+        # Node 1 should now have about 1 btc left (no dirty) and 11 (including dirty).
         assert_approx(self.nodes[1].getbalance(), 1, 0.001)
         assert_approx(self.nodes[1].getbalance(avoid_reuse=False), 11, 0.001)
 
-if __name__ == '__main__':
+if __name__ == "__main__":
```
Comment (MarcoFalke) [inline test/functional/wallet_avoidreuse.py]: Why is this changed. Most other tests use single quotes. In python both serve the same function and I don't see the need to change this.

File: test/functional/wallet_avoidreuse.py [hunk 6]
```diff
@@ -138,51 +145,52 @@ def test_fund_send_fund_senddirty(self):
...
 
     def test_fund_send_fund_send(self):
-        '''
+        """
         Test the simple case where [1] generates a new address A, then
-        [0] sends 10 BTC to A.
-        [1] spends 5 BTC from A. (leaving roughly 5 BTC useable)
-        [0] sends 10 BTC to A again.
-        [1] tries to spend 10 BTC (fails; dirty).
+        [0] sends 10 BTC to A
```
Comment (MarcoFalke) [inline test/functional/wallet_avoidreuse.py]: Here you remove the dots, elsewhere you add them?

File: test/functional/wallet_avoidreuse.py (hunk 2 above)
Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Thanks @jachiang. I removed both articles from the docstring ("that" / "the") for the log version to keep it shorter, as it was longer than the other logs, and for consistency because the other added logs don't have articles either.

**Review by maflcko — COMMENTED** [review]
Just some general notes. I don't see a reason to do most of these changes. Most look like irrelevant character shuffling, which neither the python interpreter should care about nor any human reader of the code.

No need to change them and invalidate the ACK, but something to keep in mind for the future.

File: test/functional/wallet_avoidreuse.py (hunk 4 above)
Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: I was following https://www.python.org/dev/peps/pep-0008/#comments

File: test/functional/wallet_avoidreuse.py (hunk 5 above)
Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Was being consistent with the rest of this test which is mostly in double quotes and with https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: test/functional/wallet_avoidreuse.py (hunk 6 above)
Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: This is a list of partial sentences.

**Comment by jonatack** [comment]
@MarcoFalke, thank you for reviewing. I empathise with your remarks, the increased review burden, and the issues of churn, git blame & git history. `test/functional/README.md` says "Where possible, try to adhere to [PEP-8 guidelines](https://www.python.org/dev/peps/pep-0008/)" but yes, that should mainly be applied to the changes themselves. Happy to follow whatever you prefer and minimise the changeset.

File: test/functional/wallet_avoidreuse.py (hunk 3 above)
Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Done

**Review by jonatack — COMMENTED** [review]
Rebased to minimal changeset, no code changes except the following:
```diff
-self.extra_args = [["-whitelist=127.0.0.1"]] * 2
+self.extra_args = [["-whitelist=127.0.0.1"]] * self.num_nodes
```

**Comment by MarcoFalke** [comment]
Thanks. This patch is a lot more candy to my eyes than the previous one.

**Comment by MarcoFalke** [comment]
ACK 0e7c90eb37a687158c261ddd1ff9f1028a1e7012 🐊

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 0e7c90eb37a687158c261ddd1ff9f1028a1e7012 🐊
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjgfgv/c+HAPiENUjhlN/9JBHfsmMEootFcApeIeuXMOontskood9TPTOARVNt+
jOMDycGM8zEcJh+Fb/Oe6UyI82rUKtKfP6bGGxbhZKRGV+qBIQLOijcu+HyFdbKO
WC6Xg7L8l9v0DllSkWuDx13kgGMUmSLjLGLEptguPobmeSDrHdL/0/HZH35aH7zA
1lzEly99JabUDJckc78G1XYXHWXKImfGxsZHmDwZbGqTJaygEiCa0Psab4iI4W17
VUcaRU/kGCA5owhy6n4tYeGo6/bi4oZ5ImPJ9NkhcNAQJyjV18Rt9y1h2B5T1hXE
CeqXvEvfcXMsrKk1cG9cyNY/V6DnR0zp8jEdDu8BO4LHI9Kx5DmVVFS0p5Icpmgj
w0kD2fGy0KDPDECauFhDB0VTL1x1Fh3iet+QXeW/G4fd6/4FEGYoIUWLtwN9hVwT
M4f50hLAjTyr7qKtebkirMXH2u3dhIwXosLsxK8o+QdyB8YnkMl2mHBre18bgLxU
1pv7IBZ7
=u/gW
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6cefb31533317d8b2a2f9fe6e2a75f803e92835ac7b9cf4100482a5a252a11c1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086cefb31533317d8b2a2f9fe6e2a75f803e92835ac7b9cf4100482a5a252a11c1f01024bc23cd23cf531f6743a38cd275eb8b08fff0105ce972f41d03de31d514bbdd66a795f008f1045dc42443f008da66c84ee7d8bf770083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107bf9f8b53d02cd71386cf9e61e01f94c08f1045dc42443f008891d1432705901260083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010adfe2f88ab4db6adf8b0c2f2903b827a08f1045dc42443f008495d1c4472c0bf240083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ff39dff0f1f1b1b9fff1cd3a544d14cc08f1045dc42442f008921697dd595c5a520083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Review by fanquake — APPROVED** [review]
ACK 0e7c90eb37a687158c261ddd1ff9f1028a1e7012

master (772673dfbe703a383ea05a3960a98a702288d2c6):
```bash
time test/functional/test_runner.py wallet_avoidreuse.py
wallet_avoidreuse.py | ✓ Passed  | 38 s
wallet_avoidreuse.py | ✓ Passed  | 14 s
wallet_avoidreuse.py | ✓ Passed  | 27 s
wallet_avoidreuse.py | ✓ Passed  | 22 s
```

This PR (0e7c90eb37a687158c261ddd1ff9f1028a1e7012):
```bash
time test/functional/test_runner.py wallet_avoidreuse.py
wallet_avoidreuse.py | ✓ Passed  | 6 s
wallet_avoidreuse.py | ✓ Passed  | 6 s
wallet_avoidreuse.py | ✓ Passed  | 6 s
wallet_avoidreuse.py | ✓ Passed  | 6 s
```

**Comment by jonatack** [comment]
> Thanks. This patch is a lot more candy to my eyes than the previous one.

I agree -- sorry about that brainfart.
